### PR TITLE
Doctests for Dependency Tree and more properties on Doc

### PR DESCRIPTION
**Pull request by free-variation**
Other than filling out a few doctests, this PR also simplifies the Doc data model so that all data are in the `words` attribute.  There is now a property for accessing sentences that reads the `words` list to group words into sentences.

Note that this accessor is written defensively to make no assumptions about the order of words in the `words` list.   There is some performance cost but presumably this is insignificant compared to the cost of running the Stanford statistical models.

The PR also adds new properties to Doc: `pos` and `features`.

### CODE REVIEW

**Comment by kylepjohnson** [comment]
> Note that this accessor is written defensively to make no assumptions about the order of words in the words list. 

> There is some performance cost but presumably this is insignificant compared to the cost of running the Stanford statistical models.

This is a cleaner model for me to keep in mind, which I hope will be easier for outside contribs to grok, even at the expense of a some speed. Thank you.

Let me know about the `Doc.features` name. Now might be the time to settle on something, before too much is built on top.

File: src/cltkv1/utils/data_types.py [hunk 1]
```diff
@@ -89,38 +89,73 @@ class Doc:
...
+        >>> cltk_doc = cltk_nlp.analyze(text=get_example_text("lat"))
+        >>> cltk_doc.features[:3]
+        [{'Case': 'Nom', 'Degree': 'Pos', 'Gender': 'Fem', 'Number': 'Sing'}, {'Mood': 'Ind', 'Number': 'Sing', 'Person': '3', 'Tense': 'Pres', 'VerbForm': 'Fin', 'Voice': 'Act'}, {'Case': 'Nom', 'Degree': 'Pos', 'Gender': 'Fem', 'Number': 'Sing', 'PronType': 'Ind'}]
+        """
+        all_feats = self._get_words_attribute("feats")
+
+        return [
+            {}
+            if feats == "_"
+            else dict([feature.split("=") for feature in feats.split("|")])
```
Comment (free-variation) [inline src/cltkv1/utils/data_types.py]: An issue here is that this unpacking commits the core `Doc` class to the string representation of feature bundles used by Stanfordnlp.

**Comment by free-variation** [comment]
`Doc.features` name sees all right to me -- would you prefer something different?

**Comment by kylepjohnson** [comment]
`Doc.features` is very ambiguous if one is unaware of what the method is supposed to do. In general and especially in the ML paradigm, all of the word information may be considered a "feature". What is unique about your dict here is (a) exclusively morphological and (b) structured in a data container instead of being an concatenated str. See where I'm coming from?

**Comment by free-variation** [comment]
> In general and especially in the ML paradigm, all of the word information may be considered a "feature".

I see your point.  Well, in Linguistics the term is more specialized: there are phonological, morphological, syntactic, and semantic features.  When talking of the "features of a word", I think one often understands morphosyntactic features, which I suppose is why the Stanford people have just `feats`.

Shall we specify then like this?  `Doc.morphosyntactic_features` ?

**Comment by kylepjohnson** [comment]
Crazy idea -- what if we dropped the `str` altogether and just returned the `dict` as `Doc.pos`? This seems aggressive but certainly would make by default the CLTK's "value add" of smartly processed data. 

Would we go so far as to create a new morphosyntactic data type, as @free-variation you did for dependency parses?

**Comment by kylepjohnson** [comment]
> When talking of the "features of a word", I think one often understands morphosyntactic features, which I suppose is why the Stanford people have just feats.

OK. I wasn't totally aware of this.

> Shall we specify then like this? `Doc.morphosyntactic_features` ?

I like this, but wouldn't it make sense to have parallel names between this and what's in `Doc.pos`, since it's all the same info only in different form.

* `Doc.morphology` -> `dict`
* `Doc.morphology_string` -> `str`

John totally not meaning to be a pest about what is a small issue. If you feel strongly, I defer to your judgment!

**Comment by free-variation** [comment]
If we create a data type, then I would propose going whole-hog and modeling a popular theory of morphosyntactic features, similar to what exists for phonology in the `Orthophonology` module in `cltk`.  This would provide an interesting value-add, indeed.  Shall I propose something?

**Comment by kylepjohnson** [comment]
> similar to what exists for phonology in the Orthophonology module in cltk. This would provide an interesting value-add, indeed. Shall I propose something?

I'm definitely game. The key for me would be making such a new type immediately *intuitive* to philologically trained people like myself (this is why I love the `dict` you created).

Until then, for this PR how about we settle upon:
* `Doc.morphosyntactic_features` -> `dict`
* `Doc.morphosyntactic_features_str`-> `str`

The above has the benefit of nudging the user towards the more meaningful structured dict.

**Comment by free-variation** [comment]
All right.  What does `Doc.morphosyntactic_features_str` return?  The unprocessed value that e.g. Stanford generates, like ```Case=Nom|Degree=Pos|Gender=Fem|Number=Sing``` ?

**Comment by kylepjohnson** [comment]
> All right. What does Doc.morphosyntactic_features_str return? The unprocessed value that e.g. Stanford generates, like Case=Nom|Degree=Pos|Gender=Fem|Number=Sing ?

Yes. That's assuming you care to return this, at all. With the dict, it is now redundant and less convenient.

**Comment by free-variation** [comment]
Wait, 

> I like this, but wouldn't it make sense to have parallel names between this and what's in `Doc.pos`, since it's all the same info only in different form.

In general it's not the same information, since the features don't specify the category ("part of speech") of the word.  We could take a theoretical stand and assert that category is a morpho feature, but I think it may still be useful to allow for a `Doc.pos` separately.  

Relatedly, I wonder if `Doc.pos` should return the `Word.upos`, so that instead of (in the Latin case) getting the rather opaque `A1|grn1|casA|gen2|stAM` , we just get `NOUN`.

**Comment by free-variation** [comment]
Oh, and I just realized the obvious -- the unpacking of the features string should happen in the wrapper, which knows about the representation used by Stanford.  Duh.  I'll move it now.

**Comment by kylepjohnson** [comment]
You've illuminated what has confused me.

1. I didn't realize that the pos / feature distinction is coming straight out of stanford
2. In many treebanks (like English I have used), the POS tag is just NOUN, ADV, etc.. In the Greek and Latin treebanks, though, POS contains much more information (as in your doctest, `['A1|grn1|casA|gen2|stAM', 'N3|modA|tem1|gen6|stAV', 'C1|grn1|casA|gen2|stPV']`)

Would it be wrong to have ALL of this info in one place.

Maybe related question: how come the features in the doctest look so different than POS:

``` python
>>> cltk_doc.pos[:3]
['A1|grn1|casA|gen2|stAM', 'N3|modA|tem1|gen6|stAV', 'C1|grn1|casA|gen2|stPV']

>>> cltk_doc.features[:3]
[{'Case': 'Nom', 'Degree': 'Pos', 'Gender': 'Fem', 'Number': 'Sing'}, {'Mood': 'Ind', 'Number': 'Sing', 'Person': '3', 'Tense': 'Pres', 'VerbForm': 'Fin', 'Voice': 'Act'}, {'Case': 'Nom', 'Degree': 'Pos', 'Gender': 'Fem', 'Number': 'Sing', 'PronType': 'Ind'}]
```

**Comment by free-variation** [comment]
> In many treebanks (like English I have used), the POS tag is just NOUN, ADV, etc.. In the Greek and Latin treebanks, though, POS contains much more information (as in your doctest, 

Right, that's the issue.  Since historically treebanks haven't worked hard at syncing their representations -- which themselves typically reflect some theory of grammar or some tradition -- POS can mean all sorts of things.  That's why there's `upos`, which just gives one a basic category but not much else.

Even in English we see that under the impulse of providing more detailed information, features other than mere category start appearing in the POS tagset.  So in the Penn treebank you have e.g. VBZ, which is verb; 3d person singular;present --  tense and person features have crept into the tag.

I believe that that POS string for Latin maps directly to the `feats` in the Stanford representation, except that the category feature is represented in `upos`.  See the table here: https://universaldependencies.org/tagset-conversion/la-itconll-uposf.html

So what I'm suggesting is that we let users access the idiosyncratic representation via `Word.pos` if they want it, but in the `Doc` accessors, just provide the dict of the morpho features and the `upos` via `pos.

**Comment by kylepjohnson** [comment]
> So what I'm suggesting is that we let users access the idiosyncratic representation via Word.pos if they want it, but in the Doc accessors, just provide the dict of the morpho features and the upos via `pos.

OK we're on the same page. I don't understand why conventions vary so much, but that's another story …

Ready for me to merge?

**Comment by free-variation** [comment]
Hang on, let me make the changes we just discussed.

**Comment by free-variation** [comment]
@kylepjohnson A bit of a messy problem.  Currently `Word.parent_token` is getting set to a Stanford word type, so Stanford representations are leaking into the CLTK classes.

We could:
1.  Just set this to the index of the word, perhaps renaming to `Word.parent_index`.
2.  Stash the index, and on a second pass through the Doc's words, set `parent_token` to the correct CLTK `Word`, perhaps renaming the attribute to `parent_word`.

I think I'd vote for the second option.

**Comment by kylepjohnson** [comment]
I'm fine with the second, too.

Jan 9, 2020, 19:12 by notifications@github.com:

**Comment by free-variation** [comment]
Actually let's just do 1 for now.  It's complicated: these parents are for multi-word tokens, which is tied to CoNLL files and whatnot.  I'm not sure how widely MWT are used.  So for now we'll just drop and index (sentence scoped).  

Of course none of this is document in `stanfordnlp`.

**Comment by kylepjohnson** [comment]
I see the following error: https://travis-ci.org/cltk/cltkv1/builds/635140704#L495

If I can fix it in 5-10 mins, will edit it in browser :)

**Comment by kylepjohnson** [comment]
Your PR runs fine on my local with `make test` but you can see what the server says.

Also, a bunch of new warnings our of the NLTK, which would be good to rm. I have a few more mins but don't expect to solve this right now.

**Comment by free-variation** [comment]
I know what it is, and the build server is right.  I went ahead and bit the bullet: both governor and parent now are full CLTK `Word`s, not just indices.  The trouble is the latter causes a self reference, which I need to interrupt.  Will get to it this evening.

What are the NLTK warnings?

**Comment by kylepjohnson** [comment]
> What are the NLTK warnings?

Don't worry about those, I guess. As of Aug 26th they've [been fixed](https://github.com/nltk/nltk/issues/2338), but the last release was Aug 20 ([3.4.5](https://pypi.org/project/nltk/#history)).

The following pytorch (by way of stanfordnlp) one is annoying since it comes up even when I have warnings suppressed (as with `make test`):
```
  /pytorch/aten/src/ATen/native/LegacyDefinitions.cpp:19: UserWarning: masked_fill_ received a mask with dtype torch.uint8, this behavior is now deprecated,please use a mask with dtype torch.bool instead.
```

I haven't spent the time to figure out what is causing it. I did do similar with this: https://github.com/cltk/cltkv1/blob/master/src/cltkv1/wrappers/stanford.py#L124

as in this case the output was breaking all of our doctests.

**Comment by free-variation** [comment]
Yeah, the PyTorch ones are annoying.  I've been assuming they'll go away in some update of stanfordnlp but haven't investigated.

**Comment by kylepjohnson** [comment]
Tests pass. Ready now?

**Comment by free-variation** [comment]
Yup, thanks.

**Comment by free-variation** [comment]
That pytorch warning is really annoying.  I'm trying to figure out where exactly it's coming from in Stanford, but they use masking everywhere so it's going to take a while.  If I find it I'll submit a PR to them.

**Comment by kylepjohnson** [comment]
Wonderful, thank you.

Next, John, are you thinking of working on more access methods? If not, I will try my hand at a few (`Doc.lemmata` and if I'm feeling saucy `Doc.embeddings`). But if you want to move on these or have a vision, let us know go for it.

Am wondering about your ideas for some kind of universal @dataclass to represent the entire mophosyntactic features of a word (as per discussion yesterday). Does this feel to you like a big experimental project or something we could start small on?

**Comment by free-variation** [comment]
Sure, I can take a stab at the latter.   For the former, there's a little utility `Doc._get_words_attribute`. if you find it useful.

Also, I did something last night that I forgot to signal: repackaged `data_types.py` and `exceptions.py` out of `utils` and into a new `core` package.  Because these are the core structures of the system!  I hope you don't mind.

**Comment by free-variation** [comment]
(The warning is coming from the parser.  Damn -- that bit is complex.  We'll see.)

**Comment by kylepjohnson** [comment]
> repackaged data_types.py and exceptions.py out of utils and into a new core package

You anticipated -- and solved -- something that was increasingly annoying me.

> former, there's a little utility Doc._get_words_attribute. if you find it useful.

https://github.com/cltk/cltkv1/blob/master/src/cltkv1/core/data_types.py#L109

So handy, thank you. Little things like this help me.

**Comment by free-variation** [comment]
Well, I fixed that warning and submitted a PR.  I'm not sure how active Manning's grad students are in managing the project though.

**Comment by kylepjohnson** [comment]
I took a look at it, I hope they can get around to it. A test of how much we could rely on them.

**Pull request merged**